### PR TITLE
SONARJAVA-894 RSPEC-2131 Primitive types boxing with toString method check

### PR DESCRIPTION
**Pull request by m-g-sonar**
Added check to control primitive types boxing when combined with `toString` method.

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/PrimitiveTypeBoxingWithToStringCheck.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  @Override
+  protected void onMethodFound(MethodInvocationTree mit) {
+    AbstractTypedTree abstractTypedTree = (AbstractTypedTree) ((MemberSelectExpressionTree) mit.methodSelect()).expression();
+    if (abstractTypedTree.is(Kind.NEW_CLASS) || isValueOfInvocation(abstractTypedTree)) {
+      String typeName = abstractTypedTree.getSymbolType().toString();
+      addIssue(mit, "Use \"" + typeName + ".toString\" instead.");
+    }
+  }
+
+  private boolean isValueOfInvocation(AbstractTypedTree abstractTypedTree) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/PrimitiveTypeBoxingWithToStringCheck.java]: You can maybe reuse MethodInvocationMatcher class, with the matches method.

File: java-checks/src/test/files/checks/PrimitiveTypeBoxingWithToStringCheck.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+class A {
+  void foo() {
+    int myInt = 4;
+    boolean myBoolean = true;
+
+    new Integer(myInt).toString(); // Noncompliant; creates & discards an Integer object
```
Comment (benzonico) [inline java-checks/src/test/files/checks/PrimitiveTypeBoxingWithToStringCheck.java]: Can you add the case of the constructor with String parameter?

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S2131.html [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<p>"Boxing" is the process of putting a primitive value into a primitive-wrapper object. When that's done purely to use the wrapper class' <code>toString</code> method, it's a waste of memory and cycles because those methods are <code>static</code>, and can therefore be used without a class instance. Similarly, using the <code>static</code> method <code>valueOf</code> without <code>String</code> argument of the primitive-wrapper classes should be avoided.</p>
+
+<h2>Noncompliant Code Example</h2>
+<pre>
+int myInt = 4;
+String myIntString = new Integer(myInt).toString(); // Noncompliant; creates & discards an Integer object
```
Comment (benzonico) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S2131.html]: Beware of ampersand : should be escaped : &amp;

Comment (m-g-sonar) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S2131.html]: fixed.

File: java-checks/src/main/java/org/sonar/java/checks/PrimitiveTypeBoxingWithToStringCheck.java (hunk 1 above)
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/PrimitiveTypeBoxingWithToStringCheck.java]: fixed in the body of the method.

**Comment by benzonico** [comment]
LGTM!

**Pull request closed without merging**